packages/firestore/src/local/web_storage.ts

```diff
@@ -0,0 +1,146 @@
+  start(): void;
+  shutdown(): void;
+}
+
+/**
+ * `PersistedWebStorage` uses Local Storage as the backing store for the
+ * WebStorage class.
+ *
+ * Once started, PersistedWebStorage will rewrite its contents to Local Storage
+ * every four seconds. Other clients may disregard its state after five seconds
+ * of inactivity.
+ */
+export class PersistedWebStorage implements WebStorage {
+  private localStorage: Storage;
+  private visibilityState: VisibilityState = VisibilityState.Unknown;
+  private started = false;
+
+  constructor(
+    private persistenceKey: string,
+    private instanceId: string,
+    private asyncQueue: AsyncQueue
+  ) {}
+
+  /** Returns true if LocalStorage is available in the current environment. */
+  static isAvailable(): boolean {
+    return typeof window !== 'undefined' && window.localStorage != null;
+  }
+
+  start(): void {
+    if (!PersistedWebStorage.isAvailable()) {
+      throw new FirestoreError(
+        Code.UNIMPLEMENTED,
+        'LocalStorage is not available on this platform.'
+      );
+    }
+
+    assert(!this.started, 'PersistedWebStorage already started');
+
+    this.localStorage = window.localStorage;
+    this.started = true;
+    this.persistState();
+    this.scheduleRefresh();
+  }
+
+  shutdown(): void {
+    assert(
+      this.started,
+      'PersistedWebStorage.shutdown() called when not started'
+    );
+    this.started = false;
+  }
+
+  setVisibility(visibilityState: VisibilityState): void {
+    this.visibilityState = visibilityState;
+    this.persistState();
+  }
+
+  private scheduleRefresh(): void {
+    this.asyncQueue.schedulePeriodically(() => {
+      if (this.started) {
+        this.persistState();
+      }
+      return Promise.resolve();
+    }, WEB_STORAGE_REFRESH_INTERVAL_MS);
+  }
+
+  /** Persists the entire known state. */
+  private persistState(): void {
+    assert(this.started, 'PersistedWebStorage not started');
+    debug(LOG_TAG, 'Persisting state in LocalStorage');
+    this.localStorage[
+      this.buildKey(VISIBILITY_PREFIX, this.persistenceKey, this.instanceId)
+    ] = this.buildValue({
+      visibilityState: VisibilityState[this.visibilityState]
+    });
+  }
+
+  /** Assembles a key for LocalStorage */
+  private buildKey(...elements: string[]): string {
+    elements.forEach(value => {
+      assert(value.indexOf('_') === -1, "Key element cannot contain '_'");
+    });
+
+    return elements.join('_');
+  }
+
+  /** JSON-encodes the provided value and its current update time. */
+  private buildValue(data: { [key: string]: string }): string {
+    const persistedData = Object.assign({ lastUpdateTime: Date.now() }, data);
+    return JSON.stringify(persistedData);
+  }
+}
+
+/**
+ * `NoOpWebStorage` implements the WebStorage API but neither persist data nor
+ * delivers notifications. This call should be used for non-persistence enabled
+ * clients.
+ */
+export class NoOpWebStorage implements WebStorage {
```

Comment:
I wonder if we even need this at all.

If persistence is disabled we don't need to create a window visibility listener and we don't need to create a tab notification channel either, maybe we can just not create them?

It's a little hard to know, I guess, and maybe it will be clearer once you write the code to actually consume this value, but something to keep in mind.

Reply:
I removed this for now, since I no longer call into WebStorage from SyncEngine.